aten/src/ATen/native/cuda/layer_norm_kernel.cu

```diff
@@ -0,0 +1,408 @@
+        LayerNormKernelImplInternal<scalar_t>(
+            X, gamma, beta, M, N, static_cast<scalar_t>(eps), Y, mean, rstd);
+      });
+}
+
+template <typename T>
+void LayerNormBackwardKernelImplInternal(
+    const Tensor& dY,
+    const Tensor& X,
+    const Tensor& mean,
+    const Tensor& rstd,
+    const Tensor& gamma,
+    int64_t M,
+    int64_t N,
+    Tensor* dX,
+    Tensor* dgamma,
+    Tensor* dbeta) {
+  DCHECK_EQ(dY.numel(), M * N);
+  DCHECK_EQ(X.numel(), M * N);
+  DCHECK_EQ(mean.numel(), M);
+  DCHECK_EQ(rstd.numel(), M);
+  DCHECK(!gamma.defined() || gamma.numel() == N);
+  const T* dY_data = dY.template data_ptr<T>();
+  const T* X_data = X.template data_ptr<T>();
+  const T* mean_data = mean.template data_ptr<T>();
+  const T* rstd_data = rstd.template data_ptr<T>();
+  const T* gamma_data =
+      gamma.defined() ? gamma.template data_ptr<T>() : nullptr;
+  T* dX_data = dX->defined() ? dX->template data_ptr<T>() : nullptr;
+  cudaStream_t cuda_stream = at::cuda::getCurrentCUDAStream();
+  if (dX_data != nullptr) {
+    Tensor ds = at::native::empty_like(mean);
+    Tensor db = at::native::empty_like(mean);
+    Tensor scale = at::native::empty_like(mean);
+    Tensor bias = at::native::empty_like(mean);
+    T* ds_data = ds.template data_ptr<T>();
+    T* db_data = db.template data_ptr<T>();
+    T* scale_data = scale.template data_ptr<T>();
+    T* bias_data = bias.template data_ptr<T>();
+    ComputeInternalGradientsCUDAKernel<T>
+        <<<M, kCUDABlockReduceNumThreads, 0, cuda_stream>>>(
+            N, dY_data, X_data, gamma_data, ds_data, db_data);
+    const int64_t B = (M + kCUDANumThreads - 1) / kCUDANumThreads;
+    ComputeGradientFusedParamsCUDAKernel<T>
+        <<<B, kCUDANumThreads, 0, cuda_stream>>>(
+            M,
+            N,
+            mean_data,
+            rstd_data,
+            ds_data,
+            db_data,
+            scale_data,
+            bias_data);
+    LayerNormBackwardCUDAKenrel<T><<<M, kCUDANumThreads, 0, cuda_stream>>>(
+        N,
+        dY_data,
+        X_data,
+        gamma_data,
+        rstd_data,
+        scale_data,
+        bias_data,
+        dX_data);
+  }
+  if (dgamma->defined() || dbeta->defined()) {
+    T* dgamma_data =
+        dgamma->defined() ? dgamma->template data_ptr<T>() : nullptr;
+    T* dbeta_data = dbeta->defined() ? dbeta->template data_ptr<T>() : nullptr;
+    if (M < 512) {
+      // For small batch size, do colwise reduce directly.
+      const int64_t B = (N + kCUDANumThreads - 1) / kCUDANumThreads;
+      GammaBetaBackwardSimpleCUDAKernel<T>
+          <<<B, kCUDANumThreads, 0, cuda_stream>>>(
+              M,
+              N,
+              dY_data,
+              X_data,
+              mean_data,
+              rstd_data,
+              dgamma_data,
+              dbeta_data);
+    } else {
+      const int64_t B = (N + 31) / 32;
+      GammaBetaBackwardCUDAKernel<T><<<B, dim3(32, 16), 0, cuda_stream>>>(
```

Comment:
Kernel as written would work only if blockDim.x = 2*blockDim.y, which happens to be true, but this limitation should be reflected in the launch parameters (something like `const int threads_y = 16; ...<<<B, dim3(2*threads_y, threads_y), 0, cuda_stream>>>`)

Reply:
Done.